File: agentkit/ts/src/integrations/functional.ts
/**
 * Function-level `FunctionTool` for one-off tools without an `Integration` subclass.
 * Mirrors `agentkit.runtime.integrations.functional`.
 *
 * The class-based `Integration` path remains the right shape for namespaced,
 * multi-tool bundles. This module adds a lightweight alternative for the common
 * case of "I have one async function; expose it as a tool."
 *
 *   const getWeather = FunctionTool(
 *     { description: "Return weather.", parameters: z.object({ city: z.string() }) },
 *     async ({ city }) => ({ city, tempF: 68 }),
 *   );
 *
 *   const runtime = new AgentBuilder().provider(p).tool(getWeather).build();
 */
import { z } from "zod";

import {
  providerSafeToolName,
  type ToolHandler,
  type ToolMeta,
  type ToolRegistry,
  type ToolSpec,
  toolParametersToJSONSchema,
} from "../tools/registry";

type ArgsOf<P> = P extends z.ZodType ? z.infer<P> : Record<string, unknown>;

/** Signature accepted by `FunctionTool`. The handler receives parsed arguments
 * (typed via Zod inference) directly — no `ctx` parameter. The return value is
 * normalised to an object (primitives become `{ result: <value> }`) before
 * being emitted as the tool result. */
export type FunctionToolHandler<P> = (args: ArgsOf<P>) => Promise<unknown>;

/** Tool packaged with its spec + adapter handler, registrable like an Integration. */
export class BoundTool {
  constructor(
    readonly spec: ToolSpec,
    readonly handler: ToolHandler,
    readonly namespace: string = "fn",
  ) {}

  register(registry: ToolRegistry): void {
    const catalogName = `${this.namespace}.${this.spec.name}`;
    registry.register(
      { ...this.spec, name: providerSafeToolName(catalogName), catalogName },
      this.handler,
    );
  }
}

export interface FunctionToolMeta<P> extends Omit<ToolMeta, "parameters"> {
  /** Tool name. Defaults to the handler's `.name` if available. */
  name?: string;
  /** Zod schema or JSON Schema describing the arguments. */
  parameters: P;
  /** Namespace prefix (defaults to "fn"). */
  namespace?: string;
}

/**
 * Build a single tool from a meta + handler pair. The handler receives parsed
 * arguments (typed via Zod inference) directly.
 */
export function FunctionTool<P extends z.ZodType | Record<string, unknown>>(
  meta: FunctionToolMeta<P>,
  handler: FunctionToolHandler<P>,
): BoundTool {
  const name = meta.name ?? handler.name ?? "tool";
  const spec: ToolSpec = {
    name,
    description: meta.description,
    parameters: toolParametersToJSONSchema(meta.parameters as never),
    ...(meta.risk !== undefined ? { risk: meta.risk } : {}),
    ...(meta.tags !== undefined ? { tags: meta.tags } : {}),
    ...(meta.enabled !== undefined ? { enabled: meta.enabled } : {}),
  };

  const adapter: ToolHandler = async (_ctx, args) => {
    const result = await handler(args as ArgsOf<P>);
    // ToolHandler must return an object; wrap primitives/arrays.
    if (result !== null && typeof result === "object" && !Array.isArray(result)) {
      return result as Record<string, unknown>;
    }
    return { result };
  };
  return new BoundTool(spec, adapter, meta.namespace ?? "fn");
}

File: kaji/sdk/src/runtime/integrations/functional.py
"""Function-level ``@tool`` for one-off tools without an ``Integration`` subclass.

The class-based ``Integration`` path remains the right shape for namespaced,
multi-tool bundles. This module adds a lightweight alternative for the common
case of "I have one async function; expose it as a tool."

    @function_tool(risk="read")
    async def get_weather(context: ToolExecutionContext, city: str) -> dict:
        return {"city": city, "principal": context.principal_id, "tempF": 68}

    runtime = (
        AgentBuilder()
        .provider(p)
        .tool(get_weather)
        .default_context(TurnContext(principal_id="weather-app"))
        .build()
    )

The handler's type hints are introspected to build a Pydantic model, which is
then converted to JSON Schema the same way ``@tool(parameters=Model)`` does.
"""

from __future__ import annotations

import inspect
import logging
from dataclasses import replace
from typing import Any, Callable, Dict, Optional, Type, Union, get_type_hints

from pydantic import BaseModel, ConfigDict, Field, create_model

from kaji.runtime.context import ToolExecutionContext
from kaji.runtime.tools.registry import (
    ToolHandler,
    ToolRegistry,
    ToolRisk,
    ToolSpec,
    provider_safe_tool_name,
)

logger = logging.getLogger(__name__)


def _warn_on_sanitize(original: str, sanitized: str) -> None:
    logger.warning(
        "tool name %r sanitized to %r for provider compatibility",
        original,
        sanitized,
    )


class BoundTool:
    """A single tool packaged with its spec and handler, registrable like an
    ``Integration``. Built by the ``@tool`` decorator; consumed by
    ``AgentBuilder.tool(...)``.
    """

    __slots__ = ("spec", "handler", "namespace")

    def __init__(
        self, spec: ToolSpec, handler: ToolHandler, namespace: str = "fn"
    ) -> None:
        self.spec = spec
        self.handler = handler
        self.namespace = namespace

    def register(self, registry: ToolRegistry) -> None:
        catalog_name = f"{self.namespace}.{self.spec.name}"
        prefixed = replace(
            self.spec,
            name=provider_safe_tool_name(catalog_name, on_mutate=_warn_on_sanitize),
            catalog_name=catalog_name,
        )
        registry.register(prefixed)(self.handler)


def _context_parameter(fn: Callable[..., Any]) -> str | None:
    """Find an explicit context parameter without exception-based probing."""
    parameters = [
        parameter
        for parameter in inspect.signature(fn).parameters.values()
        if parameter.name != "self"
    ]
    if not parameters:
        return None
    hints = get_type_hints(fn)
    first = parameters[0]
    if first.name == "ctx" or hints.get(first.name) is ToolExecutionContext:
        return first.name
    if any(
        hints.get(parameter.name) is ToolExecutionContext
        for parameter in parameters[1:]
    ):
        raise TypeError(
            "ToolExecutionContext must be the first function tool parameter"
        )
    return None


def _model_from_signature(
    fn: Callable[..., Any], model_name: str, context_parameter: str | None
) -> Type[BaseModel]:
    """Build a Pydantic model from the handler's annotated parameters.

    Skips ``self`` and the explicitly detected execution-context parameter.
    Raises ``TypeError`` if a tool argument has no annotation; we fail loudly
    rather than guess.
    """
    sig = inspect.signature(fn)
    hints = get_type_hints(fn)
    fields: Dict[str, Any] = {}
    for name, param in sig.parameters.items():
        if name == "self" or name == context_parameter:
            continue
        if name not in hints:
            fn_name = getattr(fn, "__name__", repr(fn))
            raise TypeError(
                f"@tool function {fn_name!r} parameter {name!r} has no type "
                "annotation; annotate it or pass parameters= explicitly."
            )
        default = param.default if param.default is not inspect.Parameter.empty else ...
        fields[name] = (hints[name], Field(default))
    return create_model(
        model_name,
        __config__=ConfigDict(extra="forbid"),
        **fields,
    )


def _wrap_handler(fn: Callable[..., Any], context_parameter: str | None) -> ToolHandler:
    """Adapt a typed function into the registry's ``(context, args)`` shape.

    The wrapper unpacks ``args_dict`` as keyword arguments and normalises the
    return value: non-dict results (primitives, lists, models) are wrapped as
    ``{"result": ...}`` so the tool log carries a JSON-serialisable object.
    """
    sig = inspect.signature(fn)
    param_names = [
        name for name in sig.parameters if name != "self" and name != context_parameter
    ]

    async def adapter(
        context: ToolExecutionContext, args: Dict[str, Any]
    ) -> Dict[str, Any]:
        kwargs = {name: args[name] for name in param_names if name in args}
        if context_parameter is not None:
            kwargs[context_parameter] = context
        result = await fn(**kwargs)
        return result if isinstance(result, dict) else {"result": result}

    return adapter


def function_tool(
    _fn: Optional[Callable[..., Any]] = None,
    *,
    description: Optional[str] = None,
    parameters: Optional[Union[Dict[str, Any], Type[BaseModel]]] = None,
    risk: Optional[ToolRisk] = None,
    tags: tuple[str, ...] = (),
    enabled: bool = True,
    namespace: str = "fn",
    parallel_safe: bool = False,
    timeout_ms: int | None = None,
) -> Any:
    """Wrap a bare async function as a registrable tool.

    Usage::

        @function_tool(risk="read")
        async def get_weather(city: str) -> dict:
            return {"city": city, "tempF": 68}

        @function_tool(description="Look up weather.", risk="read")
        async def get_weather(city: str) -> dict: ...

    When ``parameters`` is omitted, the schema is derived from the function's
    type hints. Pass ``parameters=Model`` or a JSON Schema dict to override.
    """

    def make(fn: Callable[..., Any]) -> BoundTool:
        fn_name: str = getattr(fn, "__name__", "<anonymous>")
        context_parameter = _context_parameter(fn)
        params_schema: Dict[str, Any]
        if parameters is None:
            model = _model_from_signature(fn, f"{fn_name}__Args", context_parameter)
            params_schema = model.model_json_schema(mode="validation")
        elif isinstance(parameters, type) and issubclass(parameters, BaseModel):
            params_schema = parameters.model_json_schema(mode="validation")
        else:
            params_schema = (
                parameters  # already Dict[str, Any] per the Union guard above
            )

        spec = ToolSpec(
            name=fn_name,
            description=description or getattr(fn, "__doc__", None) or fn_name,
            parameters=params_schema,
            risk=risk,
            tags=tags,
            enabled=enabled,
            parallel_safe=parallel_safe,
            timeout_ms=timeout_ms,
        )
        return BoundTool(
            spec=spec,
            handler=_wrap_handler(fn, context_parameter),
            namespace=namespace,
        )

    # Support both bare ``@tool`` and parameterised ``@tool(...)`` forms.
    if _fn is not None:
        return make(_fn)
    return make
